src/learner.cc

```diff
@@ -1031,6 +1031,18 @@ class LearnerImpl : public LearnerIO {
     std::ostringstream os;
     os << '[' << iter << ']' << std::setiosflags(std::ios::fixed);
     if (metrics_.size() == 0 && tparam_.disable_default_eval_metric <= 0) {
+      auto warn_default_eval_metric = [](const std::string& objective, const std::string& before,
```

Comment:
Can we place this warning inside objective function `DefaultEvalMetric`?  Also, are we sure this warning is only emitted once during training?

Reply:
I don't think moving the warning to `ObjFunction::DefaultEvalMetric()` is a good idea, since then the warning would appear in multiple places. I'd like to keep it here so that we can easily remove it later.

> Also, are we sure this warning is only emitted once during training?

Yes. The warning is emitted just before the default evaluation metric gets added to the vector `metrics_`. Once the default metric is in `metrics_`, the warning will not be thrown.